### PR TITLE
#382 implement authentication

### PR DESCRIPTION
**Pull request by stefanatanasovmk**

### CODE REVIEW

**Review by edichoska — CHANGES_REQUESTED** [review]
Is this tested and working? @stefanatanasovmk @MarijaTrajkoska

**Comment by edichoska** [comment]
Please resolve conflicts before requesting a review 🙏 
Also it is a good practice to run npm run build-dev to catch the errors that can occur during vercel deployment
This is the Error:
Screenshot 2024-07-23 at 15 44 12

File: utils/actions/session.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+export async function setSession(session: Session): Promise<void> {
+  const cookie = JSON.stringify(session);
+  cookies().set('session', cookie, {
+    expires: Date.now() + 1000 * 60 * 60 * 24 * 7, // 7 days
+    httpOnly: false,
+  });
+}
+
+export async function getSession(): Promise<Session | null> {
+  const session = cookies().get('session');
```
[the remark this answers is not in the document]
Comment (stefanatanasovmk) [reply utils/actions/session.ts:20]: We don't need both, removed auth.ts

File: components/login.tsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  useEffect(() => {
+    const getCsrf = () =>
+      fetch('https://staging-api.learnhub.mk/sanctum/csrf-cookie', {
+        headers: {
+          Accept: 'application/json',
+          Referer: 'learnhub.mk',
+        },
+        credentials: 'include',
+      });
+    (async () => {
```
[the remark this answers is not in the document]
Comment (stefanatanasovmk) [reply components/login.tsx]: No, removed.

File: app/context/authContext.tsx [hunk 3]
```diff
@@ -0,0 +1,175 @@
+'use client';
+
+import React, { createContext, useContext, ReactNode, useMemo } from 'react';
+import { MutationStatus, QueryStatus, useMutation, useQuery } from '@tanstack/react-query';
+import { clearSession, getSession, setSession } from '../../utils/actions/session';
+
+const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL || 'http://localhost:8000';
```
[the remark this answers is not in the document]
Comment (stefanatanasovmk) [reply app/context/authContext.tsx]: Done !

**Comment by stefanatanasovmk** [comment]
The checks pass now. Some unused components which were committed couple of days ago, I forgot to remove with the latest final changes yesterday.

File: app/context/authContext.tsx [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL || '';
+
+export type Role = 'admin' | 'content_manager' | 'member';
+
+type UserType = {
+  id: number | string;
+  is_verified: boolean;
+  email: string;
+  role: Array<Role>;
```
Comment (davorminchorov) [inline app/context/authContext.tsx:15]: This was an array initially but we changed it to be a string for now. Only one role per user will be used for the time being. It may be changed in the near future if requirements change.

Comment (edichoska) [reply app/context/authContext.tsx:15]: @stefanatanasovmk

Comment (edichoska) [reply app/context/authContext.tsx:15]: we should add role automatically, depending from where the user logs in

File: app/context/authContext.tsx [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+  const {
+    status: logoutStatus,
+    mutate: logout,
+    error: logoutError,
+  } = useMutation({
+    mutationKey: ['logout'],
+    mutationFn: async () => {
+      const session = await getSession();
+      if (!session?.token || !session.role) throw new Error('No session found');
+      await fetch(`${getUrl(session.role)}/logout`, {
```
Comment (edichoska) [inline app/context/authContext.tsx:123]: we can create a middleware that would check the session before  each request

**Comment by edichoska** [comment]
we are also missing the refresh token mechanism implemented

File: app/context/authContext.tsx (hunk 4 above)
Comment (stefanatanasovmk) [reply app/context/authContext.tsx:15]: @davorminchorov This is a `UserType`, I'll change the role to be of type `Role` and not an array.
@edichoska The user role is fetched from the backend. When login in, the `login` function expect an argument `role` because the backend have different routes for login depending of the user role. But, I assume that the user role will be assigned on registration on the backend, as there is different urls for login, I guess there will be different urls for registration depending of the user's role.

File: app/context/authContext.tsx (hunk 5 above)
Comment (stefanatanasovmk) [reply app/context/authContext.tsx:123]: @edichoska Shall we do something like here: https://github.com/laravel/breeze-next/blob/master/src/hooks/auth.js

In this example in the `useAuth` hook we pass `middleware` prop that decide if the user needs to be authenticated, and if so we check if the `getUserStatus` from the `AuthContext` is `success` and if so we don't do anything, otherwise we redirect the user to the login route. 
This can create protected routes, everywhere we need the user to be logged in, we can call `useAuth` with passed `middleware: auth` as prop.

Example:

```
export const useAuth = ({
  middleware,
  redirectIfAuthenticatedTo,
}: {
  middleware: 'auth' | 'guest';
  redirectIfAuthenticatedTo: string;
}) => {
  const context = useContext(AuthContext);
  const router = useRouter();
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  const { getUserStatus } = context;

  useEffect(() => {
    if (middleware !== 'auth') return;
    if (getUserStatus === 'pending') return;
    if (getUserStatus === 'error') {
      router.push('/login');
    }
    if (getUserStatus === 'success' && redirectIfAuthenticatedTo)
      router.push(redirectIfAuthenticatedTo);
  }, [getUserStatus, middleware, redirectIfAuthenticatedTo, router]);
  return { ...context };
};

```

This is all in the client-side and don't have anything to do with the nextjs middleware. But in the nextjs middleware, we can check to see the expiration date of the session, and refresh it if it is expiring in a certain time?

Comment (stoimenov95) [reply app/context/authContext.tsx:123]: @stefanatanasovmk, from my side, we can go with the proposed solution, and integrating server-side checks with Next.js middleware for session expiration and refresh could provide a more robust solution.

File: app/context/authContext.tsx [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+import { useRouter } from 'next/navigation';
+import { MutationStatus, QueryStatus, useMutation, useQuery } from '@tanstack/react-query';
+import { clearSession, getSession, setSession } from '../../utils/actions/session';
+import getAuthUrl from '../../utils/getAuthUrl';
+
+const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL || '';
+
+export type Role = 'admin' | 'content_manager' | 'member';
+
+export type MiddlewareType = 'auth' | 'guest';
```
Comment (edichoska) [inline app/context/authContext.tsx:13]: lets make this enums, we will have  auto-completion with enums and will reduce the risk of making mistakes.

File: app/context/authContext.tsx [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+  role: Role;
+};
+
+interface AuthContextType {
+  user?: UserType | null;
+  login: ({
+    // eslint-disable-next-line no-unused-vars
+    email, // eslint-disable-next-line no-unused-vars
+    password, // eslint-disable-next-line no-unused-vars
+    role,
```
Comment (edichoska) [inline app/context/authContext.tsx:28]: can we create an interface here? also for the return type, so if it somehow changes in the future we don't have to remember and make the changes everywhere in the code, we would just change the interface

File: app/context/authContext.tsx (hunk 6 above)
Comment (edichoska) [inline app/context/authContext.tsx:13]: I am not sure why we have the distinction here for auth and guest. Where would guest be applicable?

**Review by edichoska — CHANGES_REQUESTED** [review]

File: app/context/authContext.tsx (hunk 6 above)
Comment (stefanatanasovmk) [reply app/context/authContext.tsx:13]: Example: on the home page we want to show "Welcome {username}" , for that we would need to get the user object from the `useAuth` hook, but the home page can be viewed as guest as well, so in that case as a `middleware` value in the hook, we would set `guest` so the `useEffect` in the `useAuth` hook don't redirect the guest users to the `/login` route.

File: components/reusable-components/navigation/Navigation.tsx [hunk 8]
```diff
@@ -24,28 +25,44 @@ const Navigation = () => {
...
+            type="link"
+            buttonClass={['primaryButton', 'navigationButton', 'button']}
+            buttonText="Регистрирај се"
+          />
+          <Button
+            href="/login"
+            type="link"
+            buttonClass={['secondaryButton', 'navigationButton', 'button']}
+            buttonText="Најави се"
+          />
```
Comment (edichoska) [inline components/reusable-components/navigation/Navigation.tsx:41]: @MarijaTrajkoska can you revert the changes made in the navigation(adding the buttons)?
We will use this buttons only for the platform. So we will add them again when we are implementing the functionalities regarding the platform

**Review by edichoska — APPROVED** [review]
great work here @stefanatanasovmk & @MarijaTrajkoska 
just the small comment and you can merge this

**Comment by edichoska** [comment]
@MarijaTrajkoska @stefanatanasovmk please resolve conflicts and tag me so we can merge this as soon as possible 🙏

**Review by edichoska — APPROVED** [review]
great job @stefanatanasovmk @MarijaTrajkoska 
lets test this 🙏
